---
title: "cnidome_asr"
author: "Noemie Sierra"
date: "2/19/2020"
output: html_document
---

```{r setup, echo=F}

  knitr::opts_chunk$set(warning = FALSE, include = FALSE)
  knitr::opts_knit$set(root.dir = "~/Desktop/github/cnidome_asr")
  
  # Set up project directory
  root="/Users/noemie/Desktop/projects/cnidome"
  dir.create(file.path(root,'data', 'alignments_18'), recursive = TRUE)
  dir.create(file.path(root,'data', 'character_data'), recursive = TRUE)


  # Load packages and tools
  suppressMessages(library(phytools))
  suppressMessages(library(gsheet))
  suppressMessages(library(stringr))
  suppressMessages(library("Biostrings"))
  `%notin%`<-Negate(`%in%`)
  

```


Tree Cleanup

Removes tips from tree without sufficiently specific nomenclature and changes any species names that have been updated on WoRMS. Saves to a nexus file for use with IQTree/BEAST. The resulting tree has 908 species (904 + 4 outgroups) when excluding species without sufficiently specific nomenclature and removing tips that have been consolidated into single species on WoRMS.


```{r Tree Cleanup, echo=FALSE}
  

  # Download relative-time calibrated tree from github, save local copy
    if (file.exists(paste0(root,"/data/tree_cnid_635.tre"))==F){
      url = "https://raw.githubusercontent.com/npicciani/picciani_et_al_2018/master/Analyses/10_Time_calibration/tree_cnid_635.tre"
      dest = paste0(root,"/data/tree_cnid_635.tre")
      download.file(url,dest,method="wget")
      rm(url,dest)
    }

  # Load in tree  
    raw_tree<-read.tree(file=paste0(root,"/data/tree_cnid_635.tre"))[[1]]
  
  # Remove branches where the species name is not sufficiently descriptive
    drop<- c("_sp$", "_gen$","_nr$","^Uncultured_","^Unverified_","^Undescribed_","Orange_sea")
    tree<-drop.tip(raw_tree,grep(paste(drop,collapse="|"),raw_tree$tip.label, value=T) ,trim.internal=TRUE,subtree=FALSE,rooted=is.rooted(raw_tree),collapse.singles=T,interactive=F); rm(drop,raw_tree)
    
   # Replace the species names with most recent nomenclature from WoRMS
    # Load in the deprecated_names document
    suppressMessages(dep<-read.csv(text=(gsheet2text('https://docs.google.com/spreadsheets/d/1T3-qX9tAqZ2CCQocp8CEVprsjDayoYmffNf_8F3_pzc/edit#gid=0'))))
    dep<-dep[,1:2]; dep<-dep[dep[,2]!="",]; dep<-dep[dep[,1]!="",]
    # Replace species on the tree
    orig_Sp<-tree$tip.label
    for (i in 1:length(tree$tip.label)){
      if(tree$tip.label[i] %in% dep$fixed_tree){
        tree$tip.label[i]<-as.character(dep$accepted_name[match(tree$tip.label[i],dep$fixed_tree)])}
    }; rm(i)
  # Prune species that now appear multiple times in tree
    drop<-which(duplicated(tree$tip.label)==T)
    tree<-drop.tip(tree,drop,trim.internal=TRUE,subtree=FALSE,rooted=is.rooted(tree),collapse.singles=T,interactive=F); rm(drop)
    
    
    #### SOME OF THE SPECIES ARE THE SAME AFTER THE NAME CHANGE (CONSOLIDATED) - NEED TO GET RID OF DUPLICATES!

  # Save to nexus for BEAUti import    
    writeNexus(tree,paste0(root,"/data/pruned_tree.nex"))
    #write.tree(tree,root,"/data/pruned_tree.tre") # save as .tre if useful
    
  # Tree info:
    out<-c("Amphimedon queenslandica, Magallana gigas, Strongylocentrotus purpuratus, Trichoplax adhaerens")
    message(sprintf("%s total species: %s cnidarian and 4 outgroups \n(%s)",length(tree$tip.label),length(tree$tip.label)-4,out)); rm(out)

    
```


Gene Data Cleanup v2 - Gene Partitions

Opens each multiple sequence alignment to remove accession numbers from the headers (for BEAST to connect the tree species and character species to the sequence data). Because cnidomes can vary between sister species, sequences without sufficiently specific nomenclature are also trimmed. An exception is made for Velella_sp (which is renamed to Velella_velella), since the genus Velella is considered monotypic.

The resulting list of species is 2450 sp long - which is reduced to 908 (904 + 4 outgroups) when excluding species that do not appear on the final fixed tree or do not have sufficiently specific nomenclature, and removing tips that have been consolidated into single species on WoRMS.

```{r message=FALSE}

  genes<-c("12S","16S","18S","28S","COI")
  trim_SpH=c()
  fasta_Sp=c()
  for (u in genes){
      # Download the sequence alignment file from github, save local copies
      if (file.exists(paste0(root,"/data/alignments_18/",u,"-trimmed-unique.fasta"))==F){
        url = paste0("https://raw.githubusercontent.com/npicciani/picciani_et_al_2018/master/Analyses/3_Processed_alignments/",u,"-trimmed-unique.fasta")
        dest = paste0(root,"/data/alignments_18/",u,"-trimmed-unique.fasta")
        download.file(url,dest,method="wget") #change method to auto before pubs
        rm(url,dest)
      }

      # Load in the fasta
        file=paste0(root,"/data/alignments_18/",u,"-trimmed-unique.fasta")
        fasta<-as.matrix(read.FASTA(file,type="DNA")); rm(file)
        
      # Change the "_sp" tag on Velella species (monotypic)
        rownames(fasta)<-gsub("Velella_sp","Velella_velella",as.vector(dimnames(fasta)[[1]]))
      
      # Remove accession numbers from fasta headers
          SpH<-row.names(fasta) # Save headers with accession numbers to a vector
        accession<-c("_[A-Z]{1,2}_?\\d+.\\d","_\\d{1,4}-?\\d{4}")
        rownames(fasta)<-gsub(paste(accession,collapse="|"),"",as.vector(dimnames(fasta )[[1]])); rm(accession)
        
      # Remove species not found on the fixed tree 
        fasta<-fasta[row.names(fasta) %in% orig_Sp,]
        trim_SpH<-append(trim_SpH, SpH[row.names(fasta)  %notin% tree$tip.label]) # %notin% initialized in first block
        
      # Replace the species names with most recent nomenclature from WoRMS to match the tree
        if (sum(row.names(fasta) %in% dep$fixed_tree)>0){
          for (i in 1:length(row.names(fasta))){
            if(row.names(fasta)[i] %in% dep$fixed_tree){row.names(fasta)[i]<-as.character(dep$accepted_name[match(row.names(fasta)[i],dep$fixed_tree)])}
          }
        }
        
      # Save species remaining in file to a vector
        fasta_Sp<-append(fasta_Sp,row.names(fasta)) #overwrite the above
      
      # Save as new fasta file
        write.FASTA(fasta,paste0(root,"/data/alignments_18/",u,"-pruned.fasta"))
    
  }

  
    # Save list of species retained in gene files to a txt
    if (sort(unique(fasta_Sp))==sort(tree$tip.label)){
      write.table(sort(unique(fasta_Sp)),file=paste0(root,"/data/alignments_18/species_list.txt"),quote=F,row.names=F,col.names=F)}
    
    # Save alphabetical list of species trimmed from fasta files to a txt file
    write.table(sort(unique(trim_SpH)),file=paste0(root,"/data/alignments_18/fasta_species-trimmed.txt"),quote=F,row.names=F,col.names=F)

rm(SpH,fasta,fasta_Sp,u,i,orig_Sp,dep, trim_SpH)

```    
    
    
    
  ## IQTree pointed out that multiple pruned gene files had duplicate sequence headers (multiple sequences for the same spacies). Remove the sequence that has more gaps "-", or if they are equivalent, remove the first one.
  
```{r}
    
    for (u in genes){
    
      # Prune species that now appear multiple times in fastas
        # Load in the fasta
        file=paste0(root,"/data/alignments_18/",u,"-pruned.fasta")
        align<-as.alignment(as.matrix(read.FASTA(file,type="DNA")));rm(file)
        fasta<-as.DNAbin(align)
          
        sec<-which(duplicated(align[["nam"]])==T)
        if(length(sec)>0){ # if there are duplicate values in the file
          fir<-match(align[["nam"]][sec],align[["nam"]])
        
          for (i in length(fir):1){
            # information in seq is the number of non-indel characters (string length - "-")
            infoA<-nchar(align[["seq"]][sec[i]]) - str_count(align[["seq"]][sec[i]],pattern="-")
            infoB<-nchar(align[["seq"]][fir[i]]) - str_count(align[["seq"]][fir[i]],pattern="-")
          
            if (infoA > infoB) fasta<-fasta[-which(duplicated(row.names(fasta)))[i],] 
            else fasta<-fasta[-match(row.names(fasta)[which(duplicated(row.names(fasta)))[i]],row.names(fasta)),]
          }
          
        # Overwrite fasta
        write.FASTA(fasta, paste0(root,"/data/alignments_18/",u,"-pruned.fasta"))
        
        # Eventually figure out how to locate the accession from the original fasta (going to have to be by sequence matching b/c many of the duplicates are the result of the nomenclature update, which isn't reflected in the trimmed-unique.fasta)
      }
        
        
    }; rm(align,fir,sec,u,i,infoA,infoB,cut,accession,fasta)
  

```


Run IQTree

Use IQTree model selection to find potential models of best fit on the sequence data to narrow models used in BEAST. Models are limited to those available in BEAUti
```{r}

    # if the folder for the day does not exist, make it and set counter to 1
    # if it does exist, while there is a folder for gene_date_count, increment count until not present

    # Now count is at a fresh position regardless of erasing variables

  # Check if there is a directory for the day, (if not, create, set count to 1)
  if (dir.exists(paste0(root, "/dev/iqtree/", gsub(paste0(format(Sys.Date(), "%Y"),"-"),"",Sys.Date())))==F){
    c=1
    dir.create(paste0(root, "/dev/iqtree/", gsub(paste0(format(Sys.Date(), "%Y"),"-"),"",Sys.Date())),recursive=TRUE,showWarnings=FALSE)
  } else {
      while (file.exists(paste0(root,"/dev/iqtree/",gsub(paste0(format(Sys.Date(), "%Y"),"-"),"",Sys.Date()),"/",genes[1],"_",gsub(paste0(format(Sys.Date(), "%Y"),"-"),"",Sys.Date()),"_",c,".log"))==T) {
        c=c+1
      }
  }
  
    # Move into directory (only within this notebook chunk)
  suppressWarnings(setwd(paste0(root, "/dev/iqtree/", gsub(paste0(format(Sys.Date(), "%Y"),"-"),"",Sys.Date()))))

  # Run IQTree
  for (u in genes){
    x<-paste0("iqtree -s ",root,"/data/alignments_18/",u,"-pruned.fasta -pre ",u,"_",gsub(paste0(format(Sys.Date(), "%Y"),"-"),"",Sys.Date()),"_",c," -st DNA -nt AUTO -m TESTNEWONLY")
    system(x)
  }

  
  #Works! Now lets actually read through BEAUti/IQTree more carefully and update the settings to test more specific/useful sets of 
  
```

# Character Data
```{r}

 # Read in character data from online source
  raw_data<-read.csv(text=(gsheet2text('https://docs.google.com/spreadsheets/d/1sQASwWIdS4UZ3SRjMNn9lRMjK9atUSq4c3sxWIRSakU/edit#gid=583503745')))
    length(unique(raw_data$Species)) # number of species with character data
  # Remove species that are not in the gene species (no gene data to place them)
  data<-raw_data[which(raw_data$Species %in% gene_species==TRUE),]; rm(raw_data)
    length(unique(data$Species)) # species on tree for which we have character data
    length(gene_species) - length(unique(data$Species)) # number of species on tree for which we are missing data



```


